# Домашнее Задание

## Частные производные

**1. Найти частные производные функции $f(x,y) = x^y$**

- $f'_{x}(x_{0},y_{0})$
- $f'_{y}(x_{0},y_{0})$

в точке $x_0 = 2, y_0 = 2$

Ответ запишите с точностью до 3 знака после запятой

In [1]:
def f(x,y):
    return x**y

x0=2
y0=2
eps = 0.000001

print round((f(x0 + eps, y0) - f(x0, y0)) / eps, 3)
print round((f(x0, y0 + eps) - f(x0, y0)) / eps, 3)

4.0
2.773


Аналитический вид частных производных:
    
- $$f'_x(x,y) = yx^{y-1}$$
- $$f'_y(x,y) = x^yln x$$

**2. Найти частные производные функции $f(x,y) = \frac{2(x+y)(1 + xy)}{(1 + x^2)(1+y^2))}$**

- $f'_{x}(x_{0},y_{0})$
- $f'_{y}(x_{0},y_{0})$

в точке $x_0 = 0, y_0 = 0$

Ответ запишите с точностью до 3 знака после запятой

In [6]:
def f(x,y):
    return (2*(x + y)*(1 + x*y))/((1 + x**2)*(1 + y**2))

x0=0
y0=0
eps = 0.000001

print round((f(x0 + eps, y0) - f(x0, y0)) / eps, 3)
print round((f(x0, y0 + eps) - f(x0, y0)) / eps, 3)

2.0
2.0


Аналитический вид частных производных:
    
- $$f'_x(x,y) = \frac{2(1-x^2)}{(1+x^2)^2}$$
- $$f'_y(x,y) = \frac{2(1-y^2)}{(1+y^2)^2}$$

**3. Найти частные производные функции**

$$f(x,y,z) = \frac{5z^2sin^2(e^{0.8x})}{e^{\frac{y^2}{2}}}$$

- $f'_{x}(x_{0},y_{0}, z_{0})$
- $f'_{y}(x_{0},y_{0}, z_{0})$
- $f'_{z}(x_{0},y_{0}, z_{0})$

в точке $x_0 = 2, y_0 = 2, z_0 = 2$

Ответ запишите с точностью до 3 знака после запятой

In [7]:
import numpy as np
def f(x, y, z):
    return (5. * z**2 * (np.sin(np.exp(0.8*x)))** 2) / (np.exp(y**2 / 2.))

x0=2
y0=2
z0=2
eps = 0.000001

print round((f(x0 + eps, y0, z0) - f(x0, y0, z0)) / eps, 3)
print round((f(x0, y0 + eps, z0) - f(x0, y0, z0)) / eps, 3)
print round((f(x0, y0, z0 + eps) - f(x0, y0, z0)) / eps, 3)

-4.965
-5.106
2.553


## Производная по направлению

Можно показать, что производная от функции $f'(x, y)$ в направлении единичного вектора $\vec{l}$:

$f'_l(x_0,y_0) = \frac{\partial f}{\partial l} = f'_x(x_0,y_0)\cdot cos\alpha + f'_y(x_0,y_0)\cdot cos\beta $, 

где

- $f'_x(x_0,y_0)$ и $f'_y(x_0,y_0)$ - частные производные по $x$ и $y$,  соответственно

- $\alpha$ и $\beta$ - углы между вектором $\vec{l}$ и $x$ и $y$, соответственно.

*Геометрический смысл этой формулы интуитивно понятен - это просто сумма компонент по обеим осям.*

Напомним, угол между двумя векторами можно найти с помощью скалярного произведения:
$$\frac{u \cdot v} {||u||_2 ||v||_2}$$

**Пример.**

Найти производную функции $f(x,y) = 3$ в точке $x_0 = -1, y_0=2$

по направлению вектора $\vec{\upsilon }(1,3)$

In [8]:
def f(x, y):
    return 3

x0 = -1
y0 = 2

fx = (f(x0+eps, y0) - f(x0, y0)) / eps
fy = (f(x0, y0 + eps) - f(x0, y0)) / eps
print (fx)
print (fy)

0.0
0.0


In [9]:
alfa = np.dot([1,3], [1,0])/(np.linalg.norm([1,3])*np.linalg.norm([1,0]))
beta = np.dot([1,3], [0,1])/(np.linalg.norm([1,3])*np.linalg.norm([0,1]))

print (alfa)
print (beta)

0.316227766017
0.948683298051


In [10]:
fv = fx*alfa + fy*beta
print (fv)

0.0


Ответ: $$\frac{\partial f(-1, 2)}{\partial \upsilon} = 0$$

** 4. Найти производную функции $f(x) = x - y$ в точке $x_0 = 0, y_0=0$ по направлению:**

1. координатных осей (параллельно им);
2. вектора $\vec{a}(2;2)$;
3. вектора $\vec{b}(-1;-3)$;
4. градиента.

In [11]:
def f(x, y):
    return x - y
x0 = 0
y0 = 0

fx = (f(x0+eps, y0) - f(x0, y0)) / eps
fy = (f(x0, y0 + eps) - f(x0, y0)) / eps
print (fx)
print (fy)

1.0
-1.0


In [14]:
#1

e1 = [1, 0]
e2 = [0, 1]

alfa_e1 = np.dot(e1, [1,0])/(np.linalg.norm(e1)*np.linalg.norm([1,0]))
beta_e1 = np.dot(e1, [0,1])/(np.linalg.norm(e1)*np.linalg.norm([0,1]))

print alfa_e1
print beta_e1

print 

alfa_e2 = np.dot(e2, [1,0])/(np.linalg.norm(e2)*np.linalg.norm([1,0]))
beta_e2 = np.dot(e2, [0,1])/(np.linalg.norm(e2)*np.linalg.norm([0,1]))

print alfa_e2
print beta_e2

print 


fe1 = fx*alfa_e1 + fy*beta_e1
print fe1

fe2 = fx*alfa_e2 + fy*beta_e2
print fe2

1.0
0.0

0.0
1.0

1.0
-1.0


Ответ: 
    
- $$\frac{\partial f(0, 0)}{\partial x} = 1$$

- $$\frac{\partial f(0, 0)}{\partial y} = -1$$

In [15]:
#2

a = [2,2]

alfa_a = np.dot(a, [1,0])/(np.linalg.norm(a)*np.linalg.norm([1,0]))
beta_a = np.dot(a, [0,1])/(np.linalg.norm(a)*np.linalg.norm([0,1]))

print alfa_a
print beta_a

print 


fa = fx*alfa_a + fy*beta_a
print fa

0.707106781187
0.707106781187

0.0


Ответ: 
    
$$\frac{\partial f(0, 0)}{\partial a} = 0$$

In [21]:
#3

b = [-1,-3]

alfa_b = np.dot(b, [1,0])/(np.linalg.norm(b)*np.linalg.norm([1,0]))
beta_b = np.dot(b, [0,1])/(np.linalg.norm(b)*np.linalg.norm([0,1]))

print alfa_b
print beta_b

print 


fb = fx*alfa_b + fy*beta_b
print fb

-0.316227766017
-0.948683298051

0.632455532034


Ответ: 
    
$$\frac{\partial f(0, 0)}{\partial b} \approx  0.63$$

Точный ответ (если найти решение аналитически):
    
$$\frac{\partial f(0, 0)}{\partial b} =  \frac{2}{\sqrt{10}}$$

Напомним определение Градиента:
$${\mathrm  {grad}}\,\varphi =\nabla \varphi ={\frac  {\partial \varphi }{\partial x}}{\vec  e}_{x}+{\frac  {\partial \varphi }{\partial y}}{\vec  e}_{y}+{\frac  {\partial \varphi }{\partial z}}{\vec  e}_{z}$$

т.е. фактически это вектор из частных производных по направлениям осей.

In [22]:
#4

print fx
print fy

grad = [fx, fy]
print grad


alfa_grad = np.dot(grad, [1,0])/(np.linalg.norm(grad)*np.linalg.norm([1,0]))
beta_grad = np.dot(grad, [0,1])/(np.linalg.norm(grad)*np.linalg.norm([0,1]))

print alfa_grad
print beta_grad

print 


fgrad = fx*alfa_grad + fy*beta_grad
print fgrad

1.0
-1.0
[1.0, -1.0]
0.707106781187
-0.707106781187

1.41421356237


Ответ: 
    
$$\frac{\partial f(0, 0)}{\partial l_{grad}} \approx  1.4$$

Точный ответ (если найти решение аналитически):

$$\frac{\partial f(0, 0)}{\partial l_{grad}} =  \sqrt{2}$$